src/librustc/middle/const_eval.rs

```diff
@@ -873,6 +899,33 @@ pub fn eval_const_expr_partial<'tcx>(tcx: &ty::ctxt<'tcx>,
       ast::ExprTup(_) => {
         const_val::Tuple(e.id)
       }
+      ast::ExprIndex(ref arr, ref idx) => {
+        let arr = try!(eval_const_expr_partial(tcx, arr, None));
+        let idx = try!(eval_const_expr_partial(tcx, idx, None));
+        let idx = match idx {
+            const_int(i) if i >= 0 => i as u64,
+            const_int(_) => signal!(e, IndexNotNatural),
+            const_uint(i) => i,
+            _ => signal!(e, IndexNotInt),
+        };
+        match arr {
+            const_val::Array(ref v) if idx as usize >= v.len() => signal!(e, IndexOutOfBounds),
+            const_val::Array(v) => try!(eval_const_expr_partial(tcx, &*v[idx as usize], None)),
+            const_val::Repeat(_, n) if idx >= n => signal!(e, IndexOutOfBounds),
+            const_val::Repeat(elem, _) => try!(eval_const_expr_partial(tcx, &*elem, None)),
+            _ => signal!(e, IndexedNonVec),
+        }
+      }
+      ast::ExprVec(ref arr) => const_val::Array(arr.clone()), // FIXME: eval elements?
```

Comment:
> But what happens if no one ever indexes the element in a constant expression?

Then they get evaluated in https://github.com/oli-obk/rust/blob/const_indexing/src/librustc_trans/trans/consts.rs#L694-717